main/src/ca/uwaterloo/flix/api/lsp/Visitor.scala

```diff
@@ -0,0 +1,937 @@
+    // TODO `insideRule` is a hack, should be removed eventually. Necessary for now since SelectChannelRule don't have locations
+    val insideRule = a.accept(chan.loc) || a.accept(exp.loc)
+    if (!insideRule) { return }
+
+    c.consumeSelectChannelRule(rule)
+
+    visitExpr(chan)
+    visitExpr(exp)
+  }
+
+  private def visitFormalParam(fparam: FormalParam)(implicit a: Acceptor, c: Consumer): Unit = {
+    val FormalParam(_, _, tpe, _, loc) = fparam
+    if (!a.accept(loc)) { return }
+
+    c.consumeFormalParam(fparam)
+
+    visitType(tpe)
+  }
+
+  private def visitHandlerRule(rule: HandlerRule)(implicit a: Acceptor, c: Consumer): Unit = {
+    val HandlerRule(op, fparams, exp) = rule
+    // TODO `insideRule` is a hack, should be removed eventually. Necessary for now since HandlerRules don't have locations
+    val insideRule = a.accept(op.loc) || fparams.map(_.loc).exists(a.accept) || a.accept(exp.loc)
+    if (!insideRule) { return }
+
+    c.consumeHandlerRule(rule)
+
+    visitOpSymUse(op)
+    fparams.foreach(visitFormalParam)
+    visitExpr(exp)
+  }
+
+  private def visitOpSymUse(symUse: OpSymUse)(implicit a: Acceptor, c: Consumer): Unit = {
+    val OpSymUse(_, loc) = symUse
+    if (!a.accept(loc)) { return }
+
+    c.consumeOpSymUse(symUse)
+  }
+
+  private def visitParYieldFrag(frag: ParYieldFragment)(implicit a: Acceptor, c: Consumer): Unit = {
+    val ParYieldFragment(pat, exp, loc) = frag
+    if (!a.accept(loc)) { return }
+
+    c.consumeParYieldFragment(frag)
+
+    visitPattern(pat)
+    visitExpr(exp)
+  }
+
+  private def visitMatchRule(rule: MatchRule)(implicit a: Acceptor, c: Consumer): Unit = {
+    val MatchRule(pat, guard, exp) = rule
+    // TODO `insideRule` is a hack, should be removed eventually. Necessary for now since MatchRules don't have locations
+    val insideRule = a.accept(pat.loc) || guard.map(_.loc).exists(a.accept) || a.accept(exp.loc)
+    if (!insideRule) { return }
+
+    c.consumeMatchRule(rule)
+
+    visitPattern(pat)
+    guard.foreach(visitExpr)
+    visitExpr(exp)
+  }
+
+  private def visitTypeMatchRule(rule: TypeMatchRule)(implicit a: Acceptor, c: Consumer): Unit = {
+    val TypeMatchRule(sym, tpe, exp) = rule
+    // TODO `insideRule` is a hack, should be removed eventually. Necessary for now since TypeMatchRules don't have locations
+    val insideRule = a.accept(sym.loc) || a.accept(tpe.loc) || a.accept(exp.loc)
+    if (!insideRule) { return }
+
+    c.consumeTypeMatchRule(rule)
+
+    visitType(tpe)
+    visitExpr(exp)
+  }
+
+  private def visitType(tpe: Type)(implicit a: Acceptor, c: Consumer): Unit = {
+    if (!a.accept(tpe.loc)) { return }
+    c.consumeType(tpe)
+  }
+
+  private def visitAnnotations(anns: Annotations)(implicit a: Acceptor, c: Consumer): Unit = {
+    val Annotations(ls) = anns
+    // TODO `insideAnns` is a hack, should be removed eventually. Necessary for now since Annotations (not to be confused with Annotation) don't have locations
```

Comment:
I think this is more tricky. So I would leave the comment, but without a TODO. This is just the way it is.

Reply:
You could maybe argue that `Annotations`s do not require visting, and that only each individual annotation ought to be visited. But then that would make it impossible to do something regarding a set of annotations (as opposed to the individual) and it would create a somewhat mismatch between the visitor and the actual structure of the AST, since `ANnotations` is an abstraction in the AST at the end of the day. That's why I added both visitAnnotations and visitAnnotation. What do you think?

Reply:
I agree with your observations. For now, I would remove visitAnnotations (plural). Later, if Annotations were to have a source location itself, we could support your use case.

Reply:
As we've been discussing, I'll keep it around with no consumer and no check :)